utils/goutils.go

```diff
@@ -0,0 +1,429 @@
+	} else {
+		// Unix:
+		osRoot = "/"
+	}
+
+	// Check if the current directory includes the go.mod file. If not, check the parent directpry
+	// and so on.
+	for {
+		// If the go.mod is found the current directory, return the path.
+		exists, err := IsFileExists(filepath.Join(wd, "go.mod"))
+		if err != nil || exists {
+			return wd, err
+		}
+
+		// If this the OS root, we can stop.
+		if wd == osRoot {
+			break
+		}
+
+		// Save this path.
+		visitedPaths[wd] = true
+		// CD to the parent directory.
+		wd = filepath.Dir(wd)
+		os.Chdir(wd)
+
+		// If we already visited this directory, it means that there's a loop and we can stop.
+		if visitedPaths[wd] {
+			return "", errors.New("Could not find go.mod for project.")
+		}
+	}
+
+	return "", errors.New("Could not find go.mod for project.")
+}
+
+// Go performs password redaction from url since version 1.13.
+// Only if go version before 1.13, should manually perform password masking.
+func shouldMaskPassword() (bool, error) {
+	return compareSpecificVersionToCurVersion(shouldMask, minGoVersionForMasking)
+}
+
+// Since version go1.16 build commands (like go build and go list) no longer modify go.mod and go.sum by default.
+func automaticallyModifyMod() (bool, error) {
+	return compareSpecificVersionToCurVersion(autoModify, maxGoVersionAutomaticallyModifyMod)
+}
+
+func compareSpecificVersionToCurVersion(result *bool, comparedVersion string) (bool, error) {
+	if result == nil {
+		goVersion, err := getParsedGoVersion()
+		if err != nil {
+			return false, err
+		}
+		autoModifyBool := !goVersion.AtLeast(comparedVersion)
+		result = &autoModifyBool
+	}
+
+	return *result, nil
+}
+
+func getParsedGoVersion() (*version.Version, error) {
+	output, err := getGoVersion()
+	if err != nil {
+		return nil, err
+	}
+	// Go version output pattern is: 'go version go1.14.1 darwin/amd64'
+	// Thus should take the third element.
+	splitOutput := strings.Split(output, " ")
+	return version.NewVersion(splitOutput[2]), nil
+}
+
+func getGoVersion() (string, error) {
+	goCmd, err := NewCmd()
+	if err != nil {
+		return "", err
+	}
+	goCmd.Command = []string{"version"}
+	output, err := gofrogcmd.RunCmdOutput(goCmd)
+	return output, err
+}
+
+// Compiles all the regex once
+func prepareGlobalRegExp() error {
+	var err error
+	if protocolRegExp == nil {
+		protocolRegExp, err = initRegExp(credentialsInUrlRegexp, RemoveCredentials)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func initRegExp(regex string, execFunc func(pattern *gofrogcmd.CmdOutputPattern) (string, error)) (*gofrogcmd.CmdOutputPattern, error) {
+	regExp, err := regexp.Compile(regex)
+	if err != nil {
+		return &gofrogcmd.CmdOutputPattern{}, err
+	}
+
+	outputPattern := &gofrogcmd.CmdOutputPattern{
+		RegExp: regExp,
+	}
+
+	outputPattern.ExecFunc = execFunc
+	return outputPattern, nil
+}
+
+// Remove the credentials information from the line.
+func RemoveCredentials(pattern *gofrogcmd.CmdOutputPattern) (string, error) {
```

Comment:
It looks like this function can become provate.
Let's also check the other functions in this file.